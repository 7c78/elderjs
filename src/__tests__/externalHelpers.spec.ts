process.cwd = () => 'test';

jest.mock('path', () => ({
  join: (...strings) => strings.join('/').replace('./', '').replace('//', '/'),
}));

const settings = {
  debug: {
    automagic: true,
  },
  locations: {
    buildFolder: './___ELDER___/',
    srcFolder: './src/',
  },
};
const query = {};

class StatSyncError extends Error {
  code: 'ENOENT';

  constructor(msg: string) {
    super(msg);
    this.code = 'ENOENT';
  }
}

describe('#externalHelpers', () => {
  beforeEach(() => jest.resetModules());
  it('throws', async () => {
    jest.mock('fs', () => ({
      statSync: jest.fn(() => {
        throw new StatSyncError('no file');
      }),
    }));
    // eslint-disable-next-line global-require
    const externalHelpers = require('../externalHelpers').default;
    // @ts-ignore
    expect(await externalHelpers({ settings, query, helpers: [] })).toEqual(undefined);
    const modifiedSettings = {
      ...settings,
      debug: { automagic: false },
      locations: { ...settings.locations, buildFolder: '' },
    };
    expect(
      await externalHelpers({
        settings: modifiedSettings,
        query,
        helpers: [],
      }),
    ).toEqual(undefined);
  });
  it('works - buildHelpers', async () => {
    jest.mock(
      'test/___ELDER___/helpers/index.js',
      () => () =>
        Promise.resolve({
          userHelper: () => 'something',
        }),
      { virtual: true },
    );
    jest.mock('test/src/helpers/index.js', () => () => Promise.resolve({ srcHelper: jest.fn() }), { virtual: true });
    jest.mock('fs', () => ({
      statSync: jest
        .fn()
        .mockImplementationOnce(() => {
          throw new Error('');
        })
        .mockImplementationOnce(() => {}),
    }));
    // eslint-disable-next-line global-require
    const externalHelpers = require('../externalHelpers').default;
    // @ts-ignore
    expect(await externalHelpers({ settings, query, helpers: [] })).toMatchSnapshot();
    // from cache
    // @ts-ignore
    expect(await externalHelpers({ settings, query, helpers: [] })).toMatchSnapshot();
  });
  it('works - userHelpers', async () => {
    jest.mock('test/___ELDER___/helpers/index.js', () => () => ({ userHelper: jest.fn() }), { virtual: true });
    jest.mock('test/src/helpers/index.js', () => () => Promise.resolve({ srcHelper: jest.fn() }), { virtual: true });
    jest.mock('fs', () => ({
      statSync: jest
        .fn()
        .mockImplementationOnce(() => {})
        .mockImplementationOnce(() => {}),
    }));
    // eslint-disable-next-line global-require
    const externalHelpers = require('../externalHelpers').default;
    // @ts-ignore
    expect(await externalHelpers({ settings, query, helpers: [] })).toMatchSnapshot();
    // from cache
    // @ts-ignore
    expect(await externalHelpers({ settings, query, helpers: [] })).toMatchSnapshot();
  });
});
