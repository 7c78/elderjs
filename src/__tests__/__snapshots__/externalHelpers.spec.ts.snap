// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#externalHelpers works - buildHelpers 1`] = `
Object {
  "userHelper": [Function],
}
`;

exports[`#externalHelpers works - buildHelpers 2`] = `
Object {
  "userHelper": [Function],
}
`;

exports[`#externalHelpers works - userHelpers 1`] = `
Object {
  "srcHelper": [MockFunction],
}
`;

exports[`#externalHelpers works - userHelpers 2`] = `
Object {
  "srcHelper": [MockFunction],
}
`;
